{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data Analysis Sample Code\n",
    "\n",
    "This was a class assignment from the data science programming course at UC Berkeley. I received a modified version of a VCF file, which included a tab-delimited file with some values including key-value pair data, and was asked to parse it and find meaningful information. \n",
    "\n",
    "Skills I am trying to demonstrate are:\n",
    "* Data cleaning\n",
    "* Data exploration and validation\n",
    "* Analysis and reasoning \n",
    "* Applying functions on data frames\n",
    "* Working with key-value pairs, even if they look a little bit unconventional \n",
    "* Dealing with missing data "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Premise: Clinical disease data \n",
    "\n",
    "> Your boss comes to you Monday morning and says “I figured out our next step; we are going to pivot from an online craft store and become a data center for genetic disease information! I found **ClinVar** which is a repository that contains expert curated data, and it is free for the taking. This is a gold mine! Look at the file and tell me what gene and mutation combinations are classified as dangerous.”  \n",
    ">  \n",
    "> Make sure that you only give your boss the dangerous mutations and include:  \n",
    ">  \n",
    ">    1) Gene name  \n",
    ">    2) Mutation ID number  \n",
    ">    3) Mutation Position (chromosome & position)  \n",
    ">    4) Mutation value (reference & alternate bases)  \n",
    ">    5) Clinical severity   \n",
    ">    6) Disease that is implicated  \n",
    ">  \n",
    "> Use the instructor-modified \"clinvar_final.txt\". This file was modified to be not exactly the same as 'standard' .vcf file to test your data parsing skills. **This is a large file so do NOT upload it into your github repo!** Replace missing values in the dataframe with: 'Not_Given' and state in your answer how you define harmful mutations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "### VCF file description (Summarized from version 4.1)\n",
    "\n",
    "```\n",
    "* The VCF specification:\n",
    "\n",
    "VCF is a text file format which contains meta-information lines, a header line, and then data lines each containing information about a position in the genome. The format also can contain genotype information on samples for each position.\n",
    "\n",
    "* Fixed fields:\n",
    "\n",
    "There are 8 fixed fields per record. All data lines are **tab-delimited**. In all cases, missing values are specified with a dot (‘.’). \n",
    "\n",
    "1. CHROM - chromosome number\n",
    "2. POS - position DNA nuceleotide count (bases) along the chromosome\n",
    "3. ID - The unique identifier for each mutation\n",
    "4. REF - reference base(s)\n",
    "5. ALT - alternate base(s)\n",
    "6. FILTER - filter status\n",
    "7. QUAL - quality\n",
    "8. INFO - a semicolon-separated series of keys with values in the format: <key>=<data>\n",
    "\n",
    "```\n",
    "### Applicable INFO field specifications\n",
    "\n",
    "```\n",
    "GENEINFO = <Gene name>\n",
    "CLNSIG =  <Clinical significance>\n",
    "CLNDN = <Disease name>\n",
    "```\n",
    "\n",
    "### Sample ClinVar data (vcf file format - not exactly the same as the file to download!)\n",
    "\n",
    "```\n",
    "##fileformat=VCFv4.1\n",
    "##fileDate=2019-03-19\n",
    "##source=ClinVar\n",
    "##reference=GRCh38\t\t\t\t\t\t\t\n",
    "#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\n",
    "1\t949523\trs786201005\tC\tT\t.\t.\tGENEINFO=ISG15;CLNSIG=5\n",
    "1\t949696\trs672601345\tC\tCG\t.\t.\tGENEINFO=ISG15;CLNSIG=5;CLNDBN=Cancer\n",
    "1\t949739\trs672601312\tG\tT\t.\t.\tGENEINFO=ISG15;CLNDBN=Cancer\n",
    "1\t955597\trs115173026\tG\tT\t.\t.\tGENEINFO=AGRN;CLNSIG=2; CLNDBN=Cancer\n",
    "1\t955619\trs201073369\tG\tC\t.\t.\tGENEINFO=AGG;CLNDBN=Heart_dis \n",
    "1\t957640\trs6657048\tC\tT\t.\t.\tGENEINFO=AGG;CLNSIG=3;CLNDBN=Heart_dis \n",
    "1\t976059\trs544749044\tC\tT\t.\t.\tGENEINFO=AGG;CLNSIG=0;CLNDBN=Heart_dis \n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 4) Your code here\n",
    "\n",
    "# See the cells below for my data exploration. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Response\n",
    "\n",
    "### Data Import"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#fileformat=VCFv4.1\r\n",
      "#fileDate=2019-03-01\r\n",
      "#source=ClinVar\r\n",
      "#reference=GRCh38\r\n",
      "#ID=<Description=\"ClinVar Variation ID\">\r\n",
      "#INFO=<ID=AF_ESP,Number=1,Type=Float,Description=\"allele frequencies from GO-ESP\">\r\n",
      "#INFO=<ID=AF_EXAC,Number=1,Type=Float,Description=\"allele frequencies from ExAC\">\r\n",
      "#INFO=<ID=AF_TGP,Number=1,Type=Float,Description=\"allele frequencies from TGP\">\r\n",
      "#INFO=<ID=ALLELEID,Number=1,Type=Integer,Description=\"the ClinVar Allele ID\">\r\n",
      "#INFO=<ID=CLNDN,Number=.,Type=String,Description=\"ClinVar's preferred disease name for the concept specified by disease identifiers in CLNDISDB\">\r\n",
      "#INFO=<ID=CLNDNINCL,Number=.,Type=String,Description=\"For included Variant : ClinVar's preferred disease name for the concept specified by disease identifiers in CLNDISDB\">\r\n",
      "#INFO=<ID=CLNDISDB,Number=.,Type=String,Description=\"Tag-value pairs of disease database name and identifier, e.g. OMIM:NNNNNN\">\r\n",
      "##INFO=<ID=CLNDISDBINCL,Number=.,Type=String,Description=\"For included Variant: Tag-value pairs of disease database name and identifier, e.g. OMIM:NNNNNN\">\r\n",
      "#INFO=<ID=CLNHGVS,Number=.,Type=String,Description=\"Top-level (primary assembly, alt, or patch) HGVS expression.\">\r\n",
      "#INFO=<ID=CLNREVSTAT,Number=.,Type=String,Description=\"ClinVar review status for the Variation ID\">\r\n",
      "#INFO=<ID=CLNSIG,Number=.,Type=String,Description=\"Clinical significance for this single variant\">\r\n",
      "#INFO=<ID=CLNSIGCONF,Number=.,Type=String,Description=\"Conflicting clinical significance for this single variant\">\r\n",
      "#INFO=<ID=CLNSIGINCL,Number=.,Type=String,Description=\"Clinical significance for a haplotype or genotype that includes this variant. Reported as pairs of VariationID:clinical significance.\">\r\n",
      "#INFO=<ID=CLNVC,Number=1,Type=String,Description=\"Variant type\">\r\n",
      "#INFO=<ID=CLNVCSO,Number=1,Type=String,Description=\"Sequence Ontology id for variant type\">\r\n",
      "#INFO=<ID=CLNVI,Number=.,Type=String,Description=\"the variant's clinical sources reported as tag-value pairs of database and variant identifier\">\r\n",
      "#INFO=<ID=DBVARID,Number=.,Type=String,Description=\"nsv accessions from dbVar for the variant\">\r\n",
      "#INFO=<ID=GENEINFO,Number=1,Type=String,Description=\"Gene(s) for the variant reported as gene symbol:gene id. The gene symbol and id are delimited by a colon (:) and each pair is delimited by a vertical bar (|)\">\r\n",
      "#INFO=<ID=MC,Number=.,Type=String,Description=\"comma separated list of molecular consequence in the form of Sequence Ontology ID|molecular_consequence\">\r\n",
      "#INFO=<ID=ORIGIN,Number=.,Type=String,Description=\"Allele origin. One or more of the following values may be added: 0 - unknown; 1 - germline; 2 - somatic; 4 - inherited; 8 - paternal; 16 - maternal; 32 - de-novo; 64 - biparental; 128 - uniparental; 256 - not-tested; 512 - tested-inconclusive; 1073741824 - other\">\r\n",
      "#INFO=<ID=RS,Number=.,Type=String,Description=\"dbSNP ID (i.e. rs number)\">\r\n",
      "#INFO=<ID=SSR,Number=1,Type=Integer,Description=\"Variant Suspect Reason Codes. One or more of the following values may be added: 0 - unspecified, 1 - Paralog, 2 - byEST, 4 - oldAlign, 8 - Para_EST, 16 - 1kg_failed, 1024 - other\">\r\n",
      "CHROM\tPOS\tID\tREF\tALT\tFILTER\tQUAL\tINFO\r\n",
      "1\t1014O42\t475283\tG\tA\t.\t.\tAF_ESP=0.00546;AF_EXAC=0.00165;AF_TGP=0.00619;ALLELEID=446939;CLNDISDB=MedGen:C4015293,OMIM:616126,Orphanet:ORPHA319563;CLNDN=Immunodeficiency_38_with_basal_ganglia_calcification;CLNHGVS=NC_000001.11:g.1014042G>A;CLNREVSTAT=criteria_provided,_single_submitter;CLNSIG=Benign;CLNVC=single_nucleotide_variant;CLNVCSO=SO:0001483;GENEINFO=ISG15:9636;MC=SO:0001583|missense_variant;ORIGIN=1;RS=143888043\r\n",
      "1\t1O14122\t542074\tC\tT\t.\t.\tAF_ESP=0.00015;AF_EXAC=0.00010;ALLELEID=514926;CLNDISDB=MedGen:C4015293,OMIM:616126,Orphanet:ORPHA319563;CLNDN=Immunodeficiency_38_with_basal_ganglia_calcification;CLNHGVS=NC_000001.11:g.1014122C>T;CLNREVSTAT=criteria_provided,_single_submitter;CLNSIG=Uncertain_significance;CLNVC=single_nucleotide_variant;CLNVCSO=SO:0001483;GENEINFO=ISG15:9636;MC=SO:0001583|missense_variant;ORIGIN=1;RS=150861311\r\n"
     ]
    }
   ],
   "source": [
    "# Let's look at the first few lines of my data. Change the number until I can get a decent picture. \n",
    "\n",
    "!head -30 clinvar_final.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3\t179210507\t403908\tA\tG\t.\t.\tALLELEID=393412;CLNDISDB=MedGen:C0018553,Orphanet:ORPHA201,SNOMED_CT:58037000;CLNDN=Cowden_syndrome;CLNHGVS=NC_000003.12:g.179210507A>G;CLNREVSTAT=criteria_provided,_single_submitter;CLNSIG=Uncertain_significance;CLNVC=single_nucleotide_variant;CLNVCSO=SO:0001483;GENEINFO=PIK3CA:5290;MC=SO:0001583|missense_variant;ORIGIN=1;RS=1060500027\r\n",
      "3\t179210511\t526648\tT\tC\t.\t.\tALLELEID=519163;CLNDISDB=MedGen:C0018553,Orphanet:ORPHA201,SNOMED_CT:58037000;CLNDN=Cowden_syndrome;CLNHGVS=NC_000003.12:g.179210511T>C;CLNREVSTAT=criteria_provided,_single_submitter;CLNSIG=Likely_benign;CLNVC=single_nucleotide_variant;CLNVCSO=SO:0001483;GENEINFO=PIK3CA:5290;MC=SO:0001819|synonymous_variant;ORIGIN=1;RS=1480813252\r\n",
      "3\t179210515\t526640\tA\tC\t.\t.\tAF_EXAC=0.00002;ALLELEID=519178;CLNDISDB=MedGen:C0018553,Orphanet:ORPHA201,SNOMED_CT:58037000|MedGen:C0027672,SNOMED_CT:699346009;CLNDN=Cowden_syndrome|Hereditary_cancer-predisposing_syndrome;CLNHGVS=NC_000003.12:g.179210515A>C;CLNREVSTAT=criteria_provided,_multiple_submitters,_no_conflicts;CLNSIG=Uncertain_significance;CLNVC=single_nucleotide_variant;CLNVCSO=SO:0001483;GENEINFO=PIK3CA:5290;MC=SO:0001583|missense_variant;ORIGIN=1;RS=199563773\r\n",
      "3\t179210516\t246681\tA\tG\t.\t.\tAF_EXAC=0.00001;ALLELEID=245287;CLNDISDB=MedGen:C0018553,Orphanet:ORPHA201,SNOMED_CT:58037000;CLNDN=Cowden_syndrome;CLNHGVS=NC_000003.12:g.179210516A>G;CLNREVSTAT=criteria_provided,_single_submitter;CLNSIG=Uncertain_significance;CLNVC=single_nucleotide_variant;CLNVCSO=SO:0001483;GENEINFO=PIK3CA:5290;MC=SO:0001583|missense_variant;ORIGIN=1;RS=753879573\r\n",
      "3\t179210538\t259958\tA\tT\t.\t.\tAF_EXAC=0.00001;ALLELEID=251013;CLNDISDB=MedGen:C0018553,Orphanet:ORPHA201,SNOMED_CT:58037000|MedGen:CN169374;CLNDN=Cowden_syndrome|not_specified;CLNHGVS=NC_000003.12:g.179210538A>T;CLNREVSTAT=criteria_provided,_multiple_submitters,_no_conflicts;CLNSIG=Likely_benign;CLNVC=single_nucleotide_variant;CLNVCSO=SO:0001483;GENEINFO=PIK3CA:5290;MC=SO:0001819|synonymous_variant;ORIGIN=1;RS=201108344\r\n"
     ]
    }
   ],
   "source": [
    "# Last few rows \n",
    "!tail -5 clinvar_final.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CHROM\tPOS\tID\tREF\tALT\tFILTER\tQUAL\tINFO\r\n"
     ]
    }
   ],
   "source": [
    "# Check field names\n",
    "!head -28 clinvar_final.txt | tail -1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Observations about the input file\n",
    "\n",
    "* The first 27 rows are informative, but aren't actually the data that we want to import. I noticed that these all start with a single hashtag/pound sign, \"#\". When we import, we would want to exclude all of these lines. \n",
    "\n",
    "* Line 28 is the header. We want to grab these and assign to our pandas dataframe later.\n",
    "\n",
    "* Line 29 onward appears to be the data that we want. \n",
    "\n",
    "* The last lines of the dataset seem like they're still part of the data we want to analyze. \n",
    "\n",
    "* The last column, called INFO, has an array inside this single column (when delmiiting each line by tab) and this array contains multiple keys and values. The values in here are parsed by a semicolon, ';'. We should parse this to get more info later. The data dictionary for this exists in the first few rows of the file. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import file, then close the file. \n",
    "# Only fetch the lines that do NOT start with a pound sign, \"#\"\n",
    "with open('Genomics_Question.txt', 'r') as f:\n",
    "    lines = [l for l in f if not l.startswith('#')]\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "See how when I just open the raw file, each line is a separate item of a list.\n",
      "Also I can see \\t for each tab, etc.\n",
      "\n",
      "['1\\t949523\\trs786201005\\tC\\tT\\t.\\t.\\tGENEINFO=ISG15;CLNSIG=5\\n', '1\\t949696\\trs672601345\\tC\\tCG\\t.\\t.\\tGENEINFO=ISG15;CLNSIG=5;CLNDBN=Cancer\\n', '1\\t949739\\trs672601312\\tG\\tT\\t.\\t.\\tGENEINFO=ISG15;CLNDBN=Cancer\\n', '1\\t955597\\trs115173026\\tG\\tT\\t.\\t.\\tGENEINFO=AGRN;CLNSIG=2; CLNDBN=Cancer\\n', '1\\t955619\\trs201073369\\tG\\tC\\t.\\t.\\tGENEINFO=AGG;CLNDBN=Heart_dis \\n', '1\\t957640\\trs6657048\\tC\\tT\\t.\\t.\\tGENEINFO=AGG;CLNSIG=3;CLNDBN=Heart_dis \\n', '1\\t976059\\trs544749044\\tC\\tT\\t.\\t.\\tGENEINFO=AGG;CLNSIG=0;CLNDBN=Heart_dis ']\n",
      "\n",
      "\n",
      "\n",
      "By joining each line in the list, I get a nice, single string that is more readable.\n",
      "This is the data that we want to analyze.\n",
      "\n",
      "1\t949523\trs786201005\tC\tT\t.\t.\tGENEINFO=ISG15;CLNSIG=5\n",
      "1\t949696\trs672601345\tC\tCG\t.\t.\tGENEINFO=ISG15;CLNSIG=5;CLNDBN=Cancer\n",
      "1\t949739\trs672601312\tG\tT\t.\t.\tGENEINFO=ISG15;CLNDBN=Cancer\n",
      "1\t955597\trs115173026\tG\tT\t.\t.\tGENEINFO=AGRN;CLNSIG=2; CLNDBN=Cancer\n",
      "1\t955619\trs201073369\tG\tC\t.\t.\tGENEINFO=AGG;CLNDBN=Heart_dis \n",
      "1\t957640\trs6657048\tC\tT\t.\t.\tGENEINFO=AGG;CLNSIG=3;CLNDBN=Heart_dis \n",
      "1\t976059\trs544749044\tC\tT\t.\t.\tGENEINFO=AGG;CLNSIG=0;CLNDBN=Heart_dis \n"
     ]
    }
   ],
   "source": [
    "# 1 - see raw file\n",
    "print(\"See how when I just open the raw file, each line is a separate item of a list.\\\n",
    "\\nAlso I can see \\\\t for each tab, etc.\\n\")\n",
    "print(lines)\n",
    "print(\"\\n\\n\")\n",
    "\n",
    "print(\"By joining each line in the list, I get a nice, single string that is more readable.\\\n",
    "\\nThis is the data that we want to analyze.\\n\")\n",
    "# 2 - Join lines to produce a single string of all the data. \n",
    "print(''.join(lines))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>INFO</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>949523</td>\n",
       "      <td>rs786201005</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>GENEINFO=ISG15;CLNSIG=5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>949696</td>\n",
       "      <td>rs672601345</td>\n",
       "      <td>C</td>\n",
       "      <td>CG</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>GENEINFO=ISG15;CLNSIG=5;CLNDBN=Cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>949739</td>\n",
       "      <td>rs672601312</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>GENEINFO=ISG15;CLNDBN=Cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>955597</td>\n",
       "      <td>rs115173026</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>GENEINFO=AGRN;CLNSIG=2; CLNDBN=Cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>955619</td>\n",
       "      <td>rs201073369</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>GENEINFO=AGG;CLNDBN=Heart_dis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>957640</td>\n",
       "      <td>rs6657048</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>GENEINFO=AGG;CLNSIG=3;CLNDBN=Heart_dis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>976059</td>\n",
       "      <td>rs544749044</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>GENEINFO=AGG;CLNSIG=0;CLNDBN=Heart_dis</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  CHROM     POS           ID REF ALT FILTER QUAL  \\\n",
       "0     1  949523  rs786201005   C   T      .    .   \n",
       "1     1  949696  rs672601345   C  CG      .    .   \n",
       "2     1  949739  rs672601312   G   T      .    .   \n",
       "3     1  955597  rs115173026   G   T      .    .   \n",
       "4     1  955619  rs201073369   G   C      .    .   \n",
       "5     1  957640    rs6657048   C   T      .    .   \n",
       "6     1  976059  rs544749044   C   T      .    .   \n",
       "\n",
       "                                      INFO  \n",
       "0                  GENEINFO=ISG15;CLNSIG=5  \n",
       "1    GENEINFO=ISG15;CLNSIG=5;CLNDBN=Cancer  \n",
       "2             GENEINFO=ISG15;CLNDBN=Cancer  \n",
       "3    GENEINFO=AGRN;CLNSIG=2; CLNDBN=Cancer  \n",
       "4           GENEINFO=AGG;CLNDBN=Heart_dis   \n",
       "5  GENEINFO=AGG;CLNSIG=3;CLNDBN=Heart_dis   \n",
       "6  GENEINFO=AGG;CLNSIG=0;CLNDBN=Heart_dis   "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import io\n",
    "\n",
    "# 3 - read_csv \n",
    "pd.read_csv(\n",
    "    # from step 2. I was wondering what this did... \n",
    "    io.StringIO(''.join(lines)),  \n",
    "        # turns out, if I just pass in the data in this argument, read_csv will try to find a file with that name.\n",
    "        # So to tell it that we're not chasing down a file of this name, \n",
    "        # but rather, this string IS the file we want to parse, \n",
    "        # we use io.StringIO. \n",
    "    \n",
    "    # Assign column names. The sample data doesn't have headers. \n",
    "    names=[\"CHROM\", \"POS\", \"ID\", \"REF\", \"ALT\", \"FILTER\", \"QUAL\", \"INFO\"], \n",
    "    \n",
    "    # Assign data types \n",
    "    dtype={'CHROM': str, 'POS': str, 'ID': str, 'REF': str,  \n",
    "           'ALT': str, 'FILTER': str, 'QUAL': str, 'INFO': str}, # watch the column order \n",
    "    sep='\\t'\n",
    ")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Okay, so now let's import the real data.\n",
    "\n",
    "with open('clinvar_final.txt', 'r') as f:\n",
    "    lines = [l for l in f if not l.startswith('#')]\n",
    "\n",
    "clin = pd.read_csv(\n",
    "    # from step 2. I was wondering what this did... \n",
    "    io.StringIO(''.join(lines)),  \n",
    "        # turns out, if I just pass in the data in this argument, read_csv will try to find a file with that name.\n",
    "        # So to tell it that we're not chasing down a file of this name, \n",
    "        # but rather, this string IS the file we want to parse, \n",
    "        # we use io.StringIO. \n",
    "    \n",
    "    # No need to assign column names to the real data - the header's already in there\n",
    "    \n",
    "    # Assign data types \n",
    "    dtype={'CHROM': str, 'POS': str, 'ID': str, 'REF': str,  \n",
    "           'ALT': str, 'FILTER': str, 'QUAL': str, 'INFO': str}, # watch the column order \n",
    "    sep='\\t'\n",
    ")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>INFO</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1014O42</td>\n",
       "      <td>475283</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AF_ESP=0.00546;AF_EXAC=0.00165;AF_TGP=0.00619;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1O14122</td>\n",
       "      <td>542074</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AF_ESP=0.00015;AF_EXAC=0.00010;ALLELEID=514926...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1014143</td>\n",
       "      <td>183381</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ALLELEID=181485;CLNDISDB=MedGen:C4015293,OMIM:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>1014179</td>\n",
       "      <td>542075</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ALLELEID=514896;CLNDISDB=MedGen:C4015293,OMIM:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>1014217</td>\n",
       "      <td>475278</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AF_ESP=0.00515;AF_EXAC=0.00831;AF_TGP=0.00339;...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  CHROM      POS      ID REF ALT FILTER QUAL  \\\n",
       "0     1  1014O42  475283   G   A      .    .   \n",
       "1     1  1O14122  542074   C   T      .    .   \n",
       "2     1  1014143  183381   C   T      .    .   \n",
       "3     1  1014179  542075   C   T      .    .   \n",
       "4     1  1014217  475278   C   T      .    .   \n",
       "\n",
       "                                                INFO  \n",
       "0  AF_ESP=0.00546;AF_EXAC=0.00165;AF_TGP=0.00619;...  \n",
       "1  AF_ESP=0.00015;AF_EXAC=0.00010;ALLELEID=514926...  \n",
       "2  ALLELEID=181485;CLNDISDB=MedGen:C4015293,OMIM:...  \n",
       "3  ALLELEID=514896;CLNDISDB=MedGen:C4015293,OMIM:...  \n",
       "4  AF_ESP=0.00515;AF_EXAC=0.00831;AF_TGP=0.00339;...  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clin.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Let's save a copy of clin just in case I mess it up.\n",
    "\n",
    "import copy\n",
    "clin_backup = copy.deepcopy(clin) \n",
    "# can't remember if a df is mutable, so I'm just gonna take a safer route by doing a deep copy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>INFO</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1014O42</td>\n",
       "      <td>475283</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AF_ESP=0.00546;AF_EXAC=0.00165;AF_TGP=0.00619;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1O14122</td>\n",
       "      <td>542074</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AF_ESP=0.00015;AF_EXAC=0.00010;ALLELEID=514926...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1014143</td>\n",
       "      <td>183381</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ALLELEID=181485;CLNDISDB=MedGen:C4015293,OMIM:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>1014179</td>\n",
       "      <td>542075</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ALLELEID=514896;CLNDISDB=MedGen:C4015293,OMIM:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>1014217</td>\n",
       "      <td>475278</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AF_ESP=0.00515;AF_EXAC=0.00831;AF_TGP=0.00339;...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  CHROM      POS      ID REF ALT FILTER QUAL  \\\n",
       "0     1  1014O42  475283   G   A      .    .   \n",
       "1     1  1O14122  542074   C   T      .    .   \n",
       "2     1  1014143  183381   C   T      .    .   \n",
       "3     1  1014179  542075   C   T      .    .   \n",
       "4     1  1014217  475278   C   T      .    .   \n",
       "\n",
       "                                                INFO  \n",
       "0  AF_ESP=0.00546;AF_EXAC=0.00165;AF_TGP=0.00619;...  \n",
       "1  AF_ESP=0.00015;AF_EXAC=0.00010;ALLELEID=514926...  \n",
       "2  ALLELEID=181485;CLNDISDB=MedGen:C4015293,OMIM:...  \n",
       "3  ALLELEID=514896;CLNDISDB=MedGen:C4015293,OMIM:...  \n",
       "4  AF_ESP=0.00515;AF_EXAC=0.00831;AF_TGP=0.00339;...  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clin_backup.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###  Data Understanding & Cleaning\n",
    "\n",
    "What does our data look like?\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(102321, 8)"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clin.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 102321 entries, 0 to 102320\n",
      "Data columns (total 8 columns):\n",
      "CHROM     102321 non-null object\n",
      "POS       102321 non-null object\n",
      "ID        102321 non-null object\n",
      "REF       102321 non-null object\n",
      "ALT       102321 non-null object\n",
      "FILTER    102321 non-null object\n",
      "QUAL      102321 non-null object\n",
      "INFO      102321 non-null object\n",
      "dtypes: object(8)\n",
      "memory usage: 6.2+ MB\n"
     ]
    }
   ],
   "source": [
    "clin.info()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Okay, so we've got 102 thousand rows. Got 8 columns as expected. \n",
    "\n",
    "Let's dig deeper. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Copy paste the data dictionary here for easy reference. This will help me identify the right field types. \n",
    "\n",
    "1. CHROM - chromosome number\n",
    "2. POS - position DNA nuceleotide count (bases) along the chromosome\n",
    "3. ID - The unique identifier for each mutation\n",
    "4. REF - reference base(s)\n",
    "5. ALT - alternate base(s)\n",
    "6. FILTER - filter status\n",
    "7. QUAL - quality\n",
    "8. INFO - a semicolon-separated series of keys with values in the format: `<key>=<data>`\n",
    "\n",
    "#### Based on this information and the data dictionary from the VCF pdf file, here are my expectations.\n",
    "- CHROM should be integer\n",
    "- POS should be integer\n",
    "- ID should be unique - if each row is mutation \n",
    "- For REF, reference base: Each base must be one of A,C,G,T,N (case insensitive). Multiple bases are permitted.\n",
    "    - The value in the POS field refers to the position of the first base in the String.\n",
    "-  ALT: Comma separated list of alternate non-reference alleles. Options are base Strings made up of the bases A,C,G,T,N, (case insensitive) or an angle-bracketed ID String (```“<ID>”```) or a breakend replacement string\n",
    "- FILTER - filter status: PASS if this position has passed all filters, i.e. a call is made at this position. Otherwise, if the site has not passed all filters, a semicolon-separated list of codes for filters that fail.\n",
    "- QUAL - quality: Phred-scaled quality score\n",
    "    \n",
    "- INFO - must parse later \n",
    "- A single period means NULL actually."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[ True]\n",
      "[ True]\n"
     ]
    }
   ],
   "source": [
    "# Is every value in CHROM a number?\n",
    "print(clin[\"CHROM\"].str.isnumeric().unique())\n",
    "# And an integer at that?\n",
    "print(clin[\"CHROM\"].str.isdigit().unique())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Great. I can change it to integer if I want to. \n",
    "\n",
    "# clin.CHROM.astype(int) # this works without throwing an error\n",
    "clin.CHROM = clin.CHROM.astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    102321.000000\n",
       "mean          1.880113\n",
       "std           0.706169\n",
       "min           1.000000\n",
       "25%           1.000000\n",
       "50%           2.000000\n",
       "75%           2.000000\n",
       "max           3.000000\n",
       "Name: CHROM, dtype: float64"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clin.CHROM.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Okay, so the values here are always 1, 2, or 3. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['1', '2', '3'], dtype=object)"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Just confirm against the backup file\n",
    "clin_backup.CHROM.unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### POS - is everything a number?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([False,  True])"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Is every value in POS a number?\n",
    "clin[\"POS\"].str.isnumeric().unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>INFO</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1014O42</td>\n",
       "      <td>475283</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AF_ESP=0.00546;AF_EXAC=0.00165;AF_TGP=0.00619;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1O14122</td>\n",
       "      <td>542074</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AF_ESP=0.00015;AF_EXAC=0.00010;ALLELEID=514926...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHROM      POS      ID REF ALT FILTER QUAL  \\\n",
       "0      1  1014O42  475283   G   A      .    .   \n",
       "1      1  1O14122  542074   C   T      .    .   \n",
       "\n",
       "                                                INFO  \n",
       "0  AF_ESP=0.00546;AF_EXAC=0.00165;AF_TGP=0.00619;...  \n",
       "1  AF_ESP=0.00015;AF_EXAC=0.00010;ALLELEID=514926...  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Nope. There are some POS values that are not numeric. Let's see them.\n",
    "clin[~clin[\"POS\"].str.isnumeric()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    1014O42\n",
       "1    1O14122\n",
       "Name: POS, dtype: object"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# There is a random O in the number. Maybe it was an OCR problem. \n",
    "# I'm gonna assume that they're supposed to be 0, so replace them with 0.\n",
    "\n",
    "clin[~clin[\"POS\"].str.isnumeric()][\"POS\"].replace(\"o\", \"0\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0    1014O42\n",
      "1    1O14122\n",
      "2    1014143\n",
      "3    1014179\n",
      "4    1014217\n",
      "Name: POS, dtype: object\n",
      "b'\\xff\\xfe1\\x000\\x001\\x004\\x00O\\x004\\x002\\x00'\n",
      "b'\\xff\\xfe1\\x00O\\x001\\x004\\x001\\x002\\x002\\x00'\n",
      "b'\\xff\\xfe1\\x000\\x001\\x004\\x001\\x004\\x003\\x00'\n",
      "b'\\xff\\xfe1\\x000\\x001\\x004\\x001\\x007\\x009\\x00'\n",
      "b'\\xff\\xfe1\\x000\\x001\\x004\\x002\\x001\\x007\\x00'\n"
     ]
    }
   ],
   "source": [
    "# Nothing happened.\n",
    "# Okay so they're not a letter O. I'll come back and deal with this later.\n",
    "\n",
    "weird_pos = clin[\"POS\"][0:5]\n",
    "\n",
    "print(weird_pos)\n",
    "\n",
    "for val in weird_pos:\n",
    "    print(val.encode('utf-16'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    1.023190e+05\n",
       "mean     1.176866e+08\n",
       "std      7.567756e+07\n",
       "min      2.722230e+05\n",
       "25%      4.670155e+07\n",
       "50%      1.222615e+08\n",
       "75%      1.787078e+08\n",
       "max      2.484886e+08\n",
       "Name: POS, dtype: float64"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# I don't nkow how to replace it, but I can at least have it be described. \n",
    "# See if it sheds any light. \n",
    "\n",
    "clin[clin[\"POS\"].str.isnumeric()][\"POS\"].astype(int).describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Yeah they're numbers. Didn't get much insight out of this other than that the number of digits range from 5 to 8. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Is ID unique per row?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Is ID unique?\n",
    "\n",
    "assert len(clin[\"ID\"]) - len(clin[\"ID\"].unique()) == 0, \\\n",
    "\"The difference of the total number of values and the number of unique ID values should be 0.\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Yup - ID values are unique."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### REF and ALT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>C</td>\n",
       "      <td>CG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>G</td>\n",
       "      <td>GGCCGCTGCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>CG</td>\n",
       "      <td>GT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>CAAG</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>CGCCCGCCAGGAGAATGTCTTCAAGAAGTTCGACG</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>GT</td>\n",
       "      <td>AC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>T</td>\n",
       "      <td>TC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>CCT</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>216</th>\n",
       "      <td>TGCAGCTCAGGTGGGCGGGGAGGGGACGGGGCCGGGGCAGCTCAGG...</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246</th>\n",
       "      <td>GCCCCTGCCAGCCCAA</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   REF         ALT\n",
       "0                                                    G           A\n",
       "1                                                    C           T\n",
       "8                                                    C          CG\n",
       "9                                                    G           T\n",
       "12                                                   G           C\n",
       "13                                                   C           A\n",
       "15                                                   G  GGCCGCTGCT\n",
       "16                                                   C           G\n",
       "17                                                  CG          GT\n",
       "23                                                   A           G\n",
       "26                                                   A           T\n",
       "33                                                CAAG           C\n",
       "38                                                   T           C\n",
       "46                 CGCCCGCCAGGAGAATGTCTTCAAGAAGTTCGACG           C\n",
       "56                                                  GT          AC\n",
       "63                                                   T          TC\n",
       "65                                                 CCT           C\n",
       "98                                                   T           A\n",
       "216  TGCAGCTCAGGTGGGCGGGGAGGGGACGGGGCCGGGGCAGCTCAGG...           T\n",
       "246                                   GCCCCTGCCAGCCCAA           G"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Let's see REF and ALT.  \n",
    "clin[[\"REF\", \"ALT\"]].drop_duplicates()[:20]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are lots of unique values here. I can parse this out to see if the values here really contain just ACGATN. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "False\n",
      "True\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "# Does REF and ALT only contain ACGTAN? Otherwise, are there any unexpeced values?\n",
    "\n",
    "def unexpected_base(base):\n",
    "    for letter in base:\n",
    "        if letter not in \"ACGATN\":\n",
    "            return True # As in TRUE, there ARE unexpected values\n",
    "            continue\n",
    "        \n",
    "    return False # As in, no, there are NO unexpected values. \n",
    "    \n",
    "# Test this out\n",
    "print(unexpected_base(\"AAAAG\"))\n",
    "print(unexpected_base(\".\"))\n",
    "print(unexpected_base(\"ABACADG\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    False\n",
       "1    False\n",
       "2    False\n",
       "3    False\n",
       "4    False\n",
       "Name: REF, dtype: bool"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Let's try it across the whole column. \n",
    "\n",
    "clin[\"REF\"].apply(unexpected_base).head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>INFO</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [CHROM, POS, ID, REF, ALT, FILTER, QUAL, INFO]\n",
       "Index: []"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clin[clin[\"REF\"].apply(unexpected_base)].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Looks like every value in \"REF\" is one of ACGTN.\n",
    "\n",
    "Check the ALT as well."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>INFO</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [CHROM, POS, ID, REF, ALT, FILTER, QUAL, INFO]\n",
       "Index: []"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Does ALT have any characters other than ACGTN or a period \".\" for NULL?\n",
    "clin[(clin[\"ALT\"].apply(unexpected_base)) & ~(clin[\"ALT\"]==\".\")].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "No. So the values in ALT and REF look as expected. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### FILTER values.\n",
    "\n",
    "Learned from the data dictionary that there's some sort of filter being applied to the data. If the datapoint passed all of them, the value in this field will be a PASS. So let's take a closer look."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['.']\n"
     ]
    }
   ],
   "source": [
    "print(clin.FILTER.unique())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It's all NULL. I'm so glad I can deal with one fewer problem.\n",
    "\n",
    "Next.\n",
    "\n",
    "#### QUAL\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['.']\n"
     ]
    }
   ],
   "source": [
    "print(clin.QUAL.unique())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Also all NULL. So I'll ignore this. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Parsing and exploring INFO. \n",
    "\n",
    "INFO contains data that can be parsed and could probably tell us about the dangers.\n",
    "Let's take a look."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['AF_ESP=0.00546;AF_EXAC=0.00165;AF_TGP=0.00619;ALLELEID=446939;CLNDISDB=MedGen:C4015293,OMIM:616126,Orphanet:ORPHA319563;CLNDN=Immunodeficiency_38_with_basal_ganglia_calcification;CLNHGVS=NC_000001.11:g.1014042G>A;CLNREVSTAT=criteria_provided,_single_submitter;CLNSIG=Benign;CLNVC=single_nucleotide_variant;CLNVCSO=SO:0001483;GENEINFO=ISG15:9636;MC=SO:0001583|missense_variant;ORIGIN=1;RS=143888043',\n",
       "       'AF_ESP=0.00015;AF_EXAC=0.00010;ALLELEID=514926;CLNDISDB=MedGen:C4015293,OMIM:616126,Orphanet:ORPHA319563;CLNDN=Immunodeficiency_38_with_basal_ganglia_calcification;CLNHGVS=NC_000001.11:g.1014122C>T;CLNREVSTAT=criteria_provided,_single_submitter;CLNSIG=Uncertain_significance;CLNVC=single_nucleotide_variant;CLNVCSO=SO:0001483;GENEINFO=ISG15:9636;MC=SO:0001583|missense_variant;ORIGIN=1;RS=150861311',\n",
       "       'ALLELEID=181485;CLNDISDB=MedGen:C4015293,OMIM:616126,Orphanet:ORPHA319563;CLNDN=Immunodeficiency_38_with_basal_ganglia_calcification;CLNHGVS=NC_000001.11:g.1014143C>T;CLNREVSTAT=no_assertion_criteria_provided;CLNSIG=Pathogenic;CLNVC=single_nucleotide_variant;CLNVCSO=SO:0001483;CLNVI=OMIM_Allelic_Variant:147571.0003;GENEINFO=ISG15:9636;MC=SO:0001587|nonsense;ORIGIN=1;RS=786201005',\n",
       "       'ALLELEID=514896;CLNDISDB=MedGen:C4015293,OMIM:616126,Orphanet:ORPHA319563;CLNDN=Immunodeficiency_38_with_basal_ganglia_calcification;CLNHGVS=NC_000001.11:g.1014179C>T;CLNREVSTAT=criteria_provided,_single_submitter;CLNSIG=Uncertain_significance;CLNVC=single_nucleotide_variant;CLNVCSO=SO:0001483;GENEINFO=ISG15:9636;MC=SO:0001583|missense_variant;ORIGIN=1;RS=1553169766',\n",
       "       'AF_ESP=0.00515;AF_EXAC=0.00831;AF_TGP=0.00339;ALLELEID=446987;CLNDISDB=MedGen:C4015293,OMIM:616126,Orphanet:ORPHA319563;CLNDN=Immunodeficiency_38_with_basal_ganglia_calcification;CLNHGVS=NC_000001.11:g.1014217C>T;CLNREVSTAT=criteria_provided,_single_submitter;CLNSIG=Benign;CLNVC=single_nucleotide_variant;CLNVCSO=SO:0001483;GENEINFO=ISG15:9636;MC=SO:0001819|synonymous_variant;ORIGIN=1;RS=61766284'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clin.INFO.unique()[:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "That's unreadable. Let's parse it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['AF_ESP=0.00546',\n",
       " 'AF_EXAC=0.00165',\n",
       " 'AF_TGP=0.00619',\n",
       " 'ALLELEID=446939',\n",
       " 'CLNDISDB=MedGen:C4015293,OMIM:616126,Orphanet:ORPHA319563',\n",
       " 'CLNDN=Immunodeficiency_38_with_basal_ganglia_calcification',\n",
       " 'CLNHGVS=NC_000001.11:g.1014042G>A',\n",
       " 'CLNREVSTAT=criteria_provided,_single_submitter',\n",
       " 'CLNSIG=Benign',\n",
       " 'CLNVC=single_nucleotide_variant']"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# First stab at parsing this info. \n",
    "all_info_field = []\n",
    "\n",
    "for x in clin.INFO:\n",
    "    each_field = x.split(';')\n",
    "    \n",
    "    all_info_field.extend(each_field)\n",
    "    \n",
    "    \n",
    "all_info_field[:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Each parsed value inside INFO is similar do a dictionary. According to the data dictionary, the values here are encoded as a semicolon-separated series of short keys with optional values in the format: `<key>=<data>[,data].`\n",
    "\n",
    "There are so many types in here, though. Which ones do I care about? The instructions for the exam said, CLNSIG, CLNDN, and GENEINFO. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "| \"ClinVar Variation ID\" | \"ClinVar Variation ID\" |\n",
      "| AF_ESP | \"allele frequencies from GO-ESP\" |\n",
      "| AF_EXAC | \"allele frequencies from ExAC\" |\n",
      "| AF_TGP | \"allele frequencies from TGP\" |\n",
      "| ALLELEID | \"the ClinVar Allele ID\" |\n",
      "| CLNDN | \"ClinVar's preferred disease name for the concept specified by disease identifiers in CLNDISDB\" |\n",
      "| CLNDNINCL | \"For included Variant : ClinVar's preferred disease name for the concept specified by disease identifiers in CLNDISDB\" |\n",
      "| CLNDISDB | \"Tag-value pairs of disease database name and identifier, e.g. OMIM:NNNNNN\" |\n",
      "| CLNDISDBINCL | \"For included Variant: Tag-value pairs of disease database name and identifier, e.g. OMIM:NNNNNN\" |\n",
      "| CLNHGVS | \"Top-level (primary assembly, alt, or patch) HGVS expression.\" |\n",
      "| CLNREVSTAT | \"ClinVar review status for the Variation ID\" |\n",
      "| CLNSIG | \"Clinical significance for this single variant\" |\n",
      "| CLNSIGCONF | \"Conflicting clinical significance for this single variant\" |\n",
      "| CLNSIGINCL | \"Clinical significance for a haplotype or genotype that includes this variant. Reported as pairs of VariationID:clinical significance.\" |\n",
      "| CLNVC | \"Variant type\" |\n",
      "| CLNVCSO | \"Sequence Ontology id for variant type\" |\n",
      "| CLNVI | \"the variant's clinical sources reported as tag-value pairs of database and variant identifier\" |\n",
      "| DBVARID | \"nsv accessions from dbVar for the variant\" |\n",
      "| GENEINFO | \"Gene(s) for the variant reported as gene symbol:gene id. The gene symbol and id are delimited by a colon (:) and each pair is delimited by a vertical bar (|)\" |\n",
      "| MC | \"comma separated list of molecular consequence in the form of Sequence Ontology ID|molecular_consequence\" |\n",
      "| ORIGIN | \"Allele origin. One or more of the following values may be added: 0 - unknown; 1 - germline; 2 - somatic; 4 - inherited; 8 - paternal; 16 - maternal; 32 - de-novo; 64 - biparental; 128 - uniparental; 256 - not-tested; 512 - tested-inconclusive; 1073741824 - other\" |\n",
      "| RS | \"dbSNP ID (i.e. rs number)\" |\n",
      "| SSR | \"Variant Suspect Reason Codes. One or more of the following values may be added: 0 - unspecified, 1 - Paralog, 2 - byEST, 4 - oldAlign, 8 - Para_EST, 16 - 1kg_failed, 1024 - other\" |\n"
     ]
    }
   ],
   "source": [
    "# Anything else to care about? Let's import the additional information. Let's just look at it. What do you see?\n",
    "with open('clinvar_final.txt', 'r') as f:\n",
    "    add_line = [l for l in f if l.startswith('#')]\n",
    "\n",
    "# Is there any way to convert this into a dictionary relaly quickly?\n",
    "for line in add_line:\n",
    "    if \"<\" in line:\n",
    "        meat = line.split(\"<\")[1].replace(\">\", \"\").replace(\"\\n\", \"\").replace(\"#\", \"\")\n",
    "        idname = meat.split(\",\")[0].split(\"=\")[1]\n",
    "        desc = meat.split(\"Description=\")[1]\n",
    "        \n",
    "        print(\"|\", idname, \"|\", desc, \"|\")\n",
    "    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Okay, that's the best i can do with the time I have. Copy paste into an easier format.  \n",
    "\n",
    "\n",
    "|ID|Description|\n",
    "|---|---|\n",
    "| AF_ESP | \"allele frequencies from GO-ESP\" |\n",
    "| AF_EXAC | \"allele frequencies from ExAC\" |\n",
    "| AF_TGP | \"allele frequencies from TGP\" |\n",
    "| ALLELEID | \"the ClinVar Allele ID\" |\n",
    "| CLNDN | \"ClinVar's preferred disease name for the concept specified by disease identifiers in CLNDISDB\" |\n",
    "| CLNDNINCL | \"For included Variant : ClinVar's preferred disease name for the concept specified by disease identifiers in CLNDISDB\" |\n",
    "| CLNDISDB | \"Tag-value pairs of disease database name and identifier, e.g. OMIM:NNNNNN\" |\n",
    "| CLNDISDBINCL | \"For included Variant: Tag-value pairs of disease database name and identifier, e.g. OMIM:NNNNNN\" |\n",
    "| CLNHGVS | \"Top-level (primary assembly, alt, or patch) HGVS expression.\" |\n",
    "| CLNREVSTAT | \"ClinVar review status for the Variation ID\" |\n",
    "| CLNSIG | \"Clinical significance for this single variant\" |\n",
    "| CLNSIGCONF | \"Conflicting clinical significance for this single variant\" |\n",
    "| CLNSIGINCL | \"Clinical significance for a haplotype or genotype that includes this variant. Reported as pairs of VariationID:clinical significance.\" |\n",
    "| CLNVC | \"Variant type\" |\n",
    "| CLNVCSO | \"Sequence Ontology id for variant type\" |\n",
    "| CLNVI | \"the variant's clinical sources reported as tag-value pairs of database and variant identifier\" |\n",
    "| DBVARID | \"nsv accessions from dbVar for the variant\" |\n",
    "| GENEINFO | \"Gene(s) for the variant reported as gene symbol:gene id. The gene symbol and id are delimited by a colon (:) and each pair is delimited by a vertical bar (|)\" |\n",
    "| MC | \"comma separated list of molecular consequence in the form of Sequence Ontology ID|molecular_consequence\" |\n",
    "| ORIGIN | \"Allele origin. One or more of the following values may be added: 0 - unknown; 1 - germline; 2 - somatic; 4 - inherited; 8 - paternal; 16 - maternal; 32 - de-novo; 64 - biparental; 128 - uniparental; 256 - not-tested; 512 - tested-inconclusive; 1073741824 - other\" |\n",
    "| RS | \"dbSNP ID (i.e. rs number)\" |\n",
    "| SSR | \"Variant Suspect Reason Codes. One or more of the following values may be added: 0 - unspecified, 1 - Paralog, 2 - byEST, 4 - oldAlign, 8 - Para_EST, 16 - 1kg_failed, 1024 - other\" |\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I don't know what a lot of these mean, so I'll have to take a guess.\n",
    "\n",
    "- everything starting with AF_ seems to indicate where the data came from. \n",
    "- What's dangerous? Diseases are dangerous. CLNDISDB seems to be the disease identifier. Based on this, CLNDN and CLNDNINCL will show the ClinVar's preferred disease name. \n",
    "- CLNREVSTAT is the review status. I guess if we didn't review a mutation, then we can't know if it's dangerous or not.\n",
    "- Not sure what clinical significance is.\n",
    "- CLNVC, variant type, may have something interesting. What type is that? \n",
    "- CLNVCSO is the Sequence Ontology id. Ontology means study of cancer. So does having this ID mean that it's related to cancer?\n",
    "\n",
    "And out of these, we really care about:\n",
    "* GENEINFO = Gene name\n",
    "* CLNSIG =  Clinical significance\n",
    "* CLNDN = Disease name\n",
    "\n",
    "Okay, so let's try to parse the data and see what's in these fields."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Extracting values from INFO\n",
    "* GENEINFO = Gene name\n",
    "* CLNSIG =  Clinical significance\n",
    "* CLNDN = Disease name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    {'AF_ESP': '0.00546', 'AF_EXAC': '0.00165', 'A...\n",
       "1    {'AF_ESP': '0.00015', 'AF_EXAC': '0.00010', 'A...\n",
       "2    {'ALLELEID': '181485', 'CLNDISDB': 'MedGen:C40...\n",
       "3    {'ALLELEID': '514896', 'CLNDISDB': 'MedGen:C40...\n",
       "4    {'AF_ESP': '0.00515', 'AF_EXAC': '0.00831', 'A...\n",
       "5    {'AF_ESP': '0.40158', 'AF_EXAC': '0.37025', 'A...\n",
       "6    {'ALLELEID': '556509', 'CLNDISDB': 'MedGen:C40...\n",
       "7    {'ALLELEID': '556512', 'CLNDISDB': 'MedGen:C40...\n",
       "8    {'ALLELEID': '171289', 'CLNDISDB': 'MedGen:C40...\n",
       "9    {'AF_EXAC': '0.00001', 'ALLELEID': '171288', '...\n",
       "Name: INFO, dtype: object"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create a function that'll process each cell \n",
    "# and create a dictionary where dict[ID] = value\n",
    "\n",
    "def extract_info(cell):\n",
    "    result = {}\n",
    "    for x in cell.split(\";\"):\n",
    "        result[x.split(\"=\")[0]] = x.split(\"=\")[1]\n",
    "    return result\n",
    "\n",
    "# Top 10 peek\n",
    "clin.INFO[0:10].map(extract_info)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Benign'"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# If I want to extract just the value of CLNSIG, this is how to do it on a single cell:\n",
    "\n",
    "extract_info(clin.INFO[0])[\"CLNSIG\"]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "pandas.core.frame.DataFrame"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Okay, let's put these dictionaries into a column. \n",
    "\n",
    "extracted_info = pd.DataFrame({\"INFO_EXT\": clin.INFO.map(extract_info)})\n",
    "\n",
    "\n",
    "type(extracted_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "clin_v2 = pd.concat([clin, extracted_info], axis=1)\n",
    "clin_v2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>INFO</th>\n",
       "      <th>INFO_EXT</th>\n",
       "      <th>CLNDN</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1014O42</td>\n",
       "      <td>475283</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AF_ESP=0.00546;AF_EXAC=0.00165;AF_TGP=0.00619;...</td>\n",
       "      <td>{'AF_ESP': '0.00546', 'AF_EXAC': '0.00165', 'A...</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1O14122</td>\n",
       "      <td>542074</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AF_ESP=0.00015;AF_EXAC=0.00010;ALLELEID=514926...</td>\n",
       "      <td>{'AF_ESP': '0.00015', 'AF_EXAC': '0.00010', 'A...</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1014143</td>\n",
       "      <td>183381</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ALLELEID=181485;CLNDISDB=MedGen:C4015293,OMIM:...</td>\n",
       "      <td>{'ALLELEID': '181485', 'CLNDISDB': 'MedGen:C40...</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>1014179</td>\n",
       "      <td>542075</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ALLELEID=514896;CLNDISDB=MedGen:C4015293,OMIM:...</td>\n",
       "      <td>{'ALLELEID': '514896', 'CLNDISDB': 'MedGen:C40...</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>1014217</td>\n",
       "      <td>475278</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AF_ESP=0.00515;AF_EXAC=0.00831;AF_TGP=0.00339;...</td>\n",
       "      <td>{'AF_ESP': '0.00515', 'AF_EXAC': '0.00831', 'A...</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHROM      POS      ID REF ALT FILTER QUAL  \\\n",
       "0      1  1014O42  475283   G   A      .    .   \n",
       "1      1  1O14122  542074   C   T      .    .   \n",
       "2      1  1014143  183381   C   T      .    .   \n",
       "3      1  1014179  542075   C   T      .    .   \n",
       "4      1  1014217  475278   C   T      .    .   \n",
       "\n",
       "                                                INFO  \\\n",
       "0  AF_ESP=0.00546;AF_EXAC=0.00165;AF_TGP=0.00619;...   \n",
       "1  AF_ESP=0.00015;AF_EXAC=0.00010;ALLELEID=514926...   \n",
       "2  ALLELEID=181485;CLNDISDB=MedGen:C4015293,OMIM:...   \n",
       "3  ALLELEID=514896;CLNDISDB=MedGen:C4015293,OMIM:...   \n",
       "4  AF_ESP=0.00515;AF_EXAC=0.00831;AF_TGP=0.00339;...   \n",
       "\n",
       "                                            INFO_EXT  \\\n",
       "0  {'AF_ESP': '0.00546', 'AF_EXAC': '0.00165', 'A...   \n",
       "1  {'AF_ESP': '0.00015', 'AF_EXAC': '0.00010', 'A...   \n",
       "2  {'ALLELEID': '181485', 'CLNDISDB': 'MedGen:C40...   \n",
       "3  {'ALLELEID': '514896', 'CLNDISDB': 'MedGen:C40...   \n",
       "4  {'AF_ESP': '0.00515', 'AF_EXAC': '0.00831', 'A...   \n",
       "\n",
       "                                               CLNDN  \n",
       "0  Immunodeficiency_38_with_basal_ganglia_calcifi...  \n",
       "1  Immunodeficiency_38_with_basal_ganglia_calcifi...  \n",
       "2  Immunodeficiency_38_with_basal_ganglia_calcifi...  \n",
       "3  Immunodeficiency_38_with_basal_ganglia_calcifi...  \n",
       "4  Immunodeficiency_38_with_basal_ganglia_calcifi...  "
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Cool. Now that the INFO field has been parsed... create extra columns that the boss wants\n",
    "    # GENEINFO\n",
    "    # CLNSIG\n",
    "    # CLNDN\n",
    "\n",
    "# To extract a single value in the dictionary in INFO_EXT\n",
    "# clin_v2.INFO_EXT[0][\"CLNDN\"]\n",
    "\n",
    "clin_v2[\"CLNDN\"] = clin_v2.INFO_EXT.map(lambda x: x.get('CLNDN'))\n",
    "\n",
    "clin_v2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# That worked! So let's pull the other two pieces of additional info I care about. \n",
    "\n",
    "clin_v2[\"GENEINFO\"] = clin_v2.INFO_EXT.map(lambda x: x.get('GENEINFO'))\n",
    "clin_v2[\"CLNSIG\"] = clin_v2.INFO_EXT.map(lambda x: x.get('CLNSIG'))\n",
    "\n",
    "clin_v2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Making sure the length didn't change as a result of my transformation. \n",
    "\n",
    "assert len(clin_v2) == len(clin_backup)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENEINFO</th>\n",
       "      <th>ID</th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>CLNSIG</th>\n",
       "      <th>CLNDN</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>475283</td>\n",
       "      <td>1</td>\n",
       "      <td>1014O42</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Benign</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>542074</td>\n",
       "      <td>1</td>\n",
       "      <td>1O14122</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>183381</td>\n",
       "      <td>1</td>\n",
       "      <td>1014143</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>542075</td>\n",
       "      <td>1</td>\n",
       "      <td>1014179</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>475278</td>\n",
       "      <td>1</td>\n",
       "      <td>1014217</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Benign</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     GENEINFO      ID  CHROM      POS REF ALT                  CLNSIG  \\\n",
       "0  ISG15:9636  475283      1  1014O42   G   A                  Benign   \n",
       "1  ISG15:9636  542074      1  1O14122   C   T  Uncertain_significance   \n",
       "2  ISG15:9636  183381      1  1014143   C   T              Pathogenic   \n",
       "3  ISG15:9636  542075      1  1014179   C   T  Uncertain_significance   \n",
       "4  ISG15:9636  475278      1  1014217   C   T                  Benign   \n",
       "\n",
       "                                               CLNDN  \n",
       "0  Immunodeficiency_38_with_basal_ganglia_calcifi...  \n",
       "1  Immunodeficiency_38_with_basal_ganglia_calcifi...  \n",
       "2  Immunodeficiency_38_with_basal_ganglia_calcifi...  \n",
       "3  Immunodeficiency_38_with_basal_ganglia_calcifi...  \n",
       "4  Immunodeficiency_38_with_basal_ganglia_calcifi...  "
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Here are the fields we want in the end.\n",
    "clin_v2[[\"GENEINFO\", \"ID\", \"CHROM\", \"POS\", \"REF\", \"ALT\", \"CLNSIG\", \"CLNDN\"]].head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "102321\n"
     ]
    }
   ],
   "source": [
    "print(len(clin_v2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CHROM       102321\n",
       "POS         102321\n",
       "ID          102321\n",
       "REF         102321\n",
       "ALT         102321\n",
       "FILTER      102321\n",
       "QUAL        102321\n",
       "INFO        102321\n",
       "INFO_EXT    102321\n",
       "CLNDN        89651\n",
       "GENEINFO    102307\n",
       "CLNSIG      100524\n",
       "dtype: int64"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# There may be missing values.\n",
    "\n",
    "clin_v2.apply('count')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Turns out, there are some values missing in CLNDN, GENEINFO, and CLNSIG. \n",
    "\n",
    "Use fillna to fill it with 'Not_Given'. \n",
    "\n",
    "And let's overwrite the v2 df with a version that has all the transformations, so it's easier to work with."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "clin_v2 = clin_v2[[\"GENEINFO\", \"ID\", \"CHROM\", \"POS\", \"REF\", \"ALT\", \"CLNSIG\", \"CLNDN\"]].fillna('Not_Given')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENEINFO</th>\n",
       "      <th>ID</th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>CLNSIG</th>\n",
       "      <th>CLNDN</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>475283</td>\n",
       "      <td>1</td>\n",
       "      <td>1014O42</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Benign</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>542074</td>\n",
       "      <td>1</td>\n",
       "      <td>1O14122</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>183381</td>\n",
       "      <td>1</td>\n",
       "      <td>1014143</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>542075</td>\n",
       "      <td>1</td>\n",
       "      <td>1014179</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>475278</td>\n",
       "      <td>1</td>\n",
       "      <td>1014217</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Benign</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     GENEINFO      ID  CHROM      POS REF ALT                  CLNSIG  \\\n",
       "0  ISG15:9636  475283      1  1014O42   G   A                  Benign   \n",
       "1  ISG15:9636  542074      1  1O14122   C   T  Uncertain_significance   \n",
       "2  ISG15:9636  183381      1  1014143   C   T              Pathogenic   \n",
       "3  ISG15:9636  542075      1  1014179   C   T  Uncertain_significance   \n",
       "4  ISG15:9636  475278      1  1014217   C   T                  Benign   \n",
       "\n",
       "                                               CLNDN  \n",
       "0  Immunodeficiency_38_with_basal_ganglia_calcifi...  \n",
       "1  Immunodeficiency_38_with_basal_ganglia_calcifi...  \n",
       "2  Immunodeficiency_38_with_basal_ganglia_calcifi...  \n",
       "3  Immunodeficiency_38_with_basal_ganglia_calcifi...  \n",
       "4  Immunodeficiency_38_with_basal_ganglia_calcifi...  "
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# This is what I'll filter now.\n",
    "\n",
    "clin_v2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Affects',\n",
       " 'Affects,_risk_factor',\n",
       " 'Benign',\n",
       " 'Benign,_drug_response',\n",
       " 'Benign,_other',\n",
       " 'Benign/Likely_benign',\n",
       " 'Benign/Likely_benign,_association',\n",
       " 'Benign/Likely_benign,_other',\n",
       " 'Benign/Likely_benign,_protective',\n",
       " 'Benign/Likely_benign,_risk_factor',\n",
       " 'Conflicting_interpretations_of_pathogenicity',\n",
       " 'Conflicting_interpretations_of_pathogenicity,_Affects',\n",
       " 'Conflicting_interpretations_of_pathogenicity,_Affects,_association,_drug_response,_other',\n",
       " 'Conflicting_interpretations_of_pathogenicity,_Affects,_other',\n",
       " 'Conflicting_interpretations_of_pathogenicity,_association',\n",
       " 'Conflicting_interpretations_of_pathogenicity,_other',\n",
       " 'Conflicting_interpretations_of_pathogenicity,_other,_risk_factor',\n",
       " 'Conflicting_interpretations_of_pathogenicity,_protective',\n",
       " 'Conflicting_interpretations_of_pathogenicity,_risk_factor',\n",
       " 'Likely_benign',\n",
       " 'Likely_benign,_other',\n",
       " 'Likely_pathogenic',\n",
       " 'Likely_pathogenic,_association',\n",
       " 'Likely_pathogenic,_other',\n",
       " 'Likely_pathogenic,_risk_factor',\n",
       " 'Not_Given',\n",
       " 'Pathogenic',\n",
       " 'Pathogenic,_Affects',\n",
       " 'Pathogenic,_association,_protective',\n",
       " 'Pathogenic,_other',\n",
       " 'Pathogenic,_protective',\n",
       " 'Pathogenic,_risk_factor',\n",
       " 'Pathogenic/Likely_pathogenic',\n",
       " 'Pathogenic/Likely_pathogenic,_other',\n",
       " 'Pathogenic/Likely_pathogenic,_risk_factor',\n",
       " 'Uncertain_significance',\n",
       " 'Uncertain_significance,_risk_factor',\n",
       " 'association',\n",
       " 'drug_response',\n",
       " 'other',\n",
       " 'protective',\n",
       " 'protective,_risk_factor',\n",
       " 'risk_factor']"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sorted(clin_v2.CLNSIG.unique())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I did a quick google search .\n",
    "\n",
    "* **pathogenic variant**: A genetic alteration that increases an individual’s susceptibility or predisposition to a certain disease or disorder. When such a variant (or mutation) is inherited, development of symptoms is more likely, but not certain. Also called deleterious mutation, disease-causing mutation, predisposing mutation, and susceptibility gene mutation.\n",
    "\n",
    "Can't find out what drug_response, protective, affects, and risk_factor mean. So I'll focus on the characteristic of \"pathogenic\".\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "What I deem dangerous:\n",
      "['Pathogenic' 'Likely_pathogenic' 'Pathogenic/Likely_pathogenic'\n",
      " 'Pathogenic,_risk_factor' 'Pathogenic/Likely_pathogenic,_risk_factor'\n",
      " 'Likely_pathogenic,_risk_factor' 'Pathogenic/Likely_pathogenic,_other'\n",
      " 'Pathogenic,_other' 'Pathogenic,_association,_protective'\n",
      " 'Likely_pathogenic,_association' 'Pathogenic,_protective'\n",
      " 'Pathogenic,_Affects' 'Likely_pathogenic,_other']\n"
     ]
    }
   ],
   "source": [
    "# Look for clinical significance that contains \"pathogenic\", but ignore conflicting interpretations. \n",
    "# Let's see if my filter works. \n",
    "\n",
    "potentially_dangerous = \\\n",
    "    clin_v2[clin_v2.CLNSIG.str.contains(\"Pathogenic\", case=False) \n",
    "         & ~clin_v2.CLNSIG.str.contains(\"Conflicting_interpretations\", case=False)\n",
    "        ]\n",
    "\n",
    "print(\"What I deem dangerous:\")\n",
    "print(potentially_dangerous.CLNSIG.unique())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Affects', 'Affects,_risk_factor', 'Benign',\n",
       "       'Benign,_drug_response', 'Benign,_other', 'Benign/Likely_benign',\n",
       "       'Benign/Likely_benign,_association', 'Benign/Likely_benign,_other',\n",
       "       'Benign/Likely_benign,_protective',\n",
       "       'Benign/Likely_benign,_risk_factor',\n",
       "       'Conflicting_interpretations_of_pathogenicity',\n",
       "       'Conflicting_interpretations_of_pathogenicity,_Affects',\n",
       "       'Conflicting_interpretations_of_pathogenicity,_Affects,_association,_drug_response,_other',\n",
       "       'Conflicting_interpretations_of_pathogenicity,_Affects,_other',\n",
       "       'Conflicting_interpretations_of_pathogenicity,_association',\n",
       "       'Conflicting_interpretations_of_pathogenicity,_other',\n",
       "       'Conflicting_interpretations_of_pathogenicity,_other,_risk_factor',\n",
       "       'Conflicting_interpretations_of_pathogenicity,_protective',\n",
       "       'Conflicting_interpretations_of_pathogenicity,_risk_factor',\n",
       "       'Likely_benign', 'Likely_benign,_other', 'Not_Given',\n",
       "       'Uncertain_significance', 'Uncertain_significance,_risk_factor',\n",
       "       'association', 'drug_response', 'other', 'protective',\n",
       "       'protective,_risk_factor', 'risk_factor'], dtype=object)"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Let's see just those rows that I think are NOT THAT dangerous.\n",
    "\n",
    "clin_v2[~(clin_v2.CLNSIG.str.contains(\"Pathogenic\", case=False) \n",
    "        & ~clin_v2.CLNSIG.str.contains(\"Conflicting_interpretations\", case=False))\n",
    "       ][\"CLNSIG\"].sort_values().unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENEINFO</th>\n",
       "      <th>ID</th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>CLNSIG</th>\n",
       "      <th>CLNDN</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>475283</td>\n",
       "      <td>1</td>\n",
       "      <td>1014O42</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Benign</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>542074</td>\n",
       "      <td>1</td>\n",
       "      <td>1O14122</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>542075</td>\n",
       "      <td>1</td>\n",
       "      <td>1014179</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>475278</td>\n",
       "      <td>1</td>\n",
       "      <td>1014217</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Benign</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>402986</td>\n",
       "      <td>1</td>\n",
       "      <td>1014228</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Benign</td>\n",
       "      <td>not_specified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>571208</td>\n",
       "      <td>1</td>\n",
       "      <td>1014255</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>568195</td>\n",
       "      <td>1</td>\n",
       "      <td>1014276</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>542076</td>\n",
       "      <td>1</td>\n",
       "      <td>1014401</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Likely_benign</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>475281</td>\n",
       "      <td>1</td>\n",
       "      <td>1014451</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Benign</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>475282</td>\n",
       "      <td>1</td>\n",
       "      <td>1014471</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>Likely_benign</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      GENEINFO      ID  CHROM      POS REF ALT                  CLNSIG  \\\n",
       "0   ISG15:9636  475283      1  1014O42   G   A                  Benign   \n",
       "1   ISG15:9636  542074      1  1O14122   C   T  Uncertain_significance   \n",
       "3   ISG15:9636  542075      1  1014179   C   T  Uncertain_significance   \n",
       "4   ISG15:9636  475278      1  1014217   C   T                  Benign   \n",
       "5   ISG15:9636  402986      1  1014228   G   A                  Benign   \n",
       "6   ISG15:9636  571208      1  1014255   G   A  Uncertain_significance   \n",
       "7   ISG15:9636  568195      1  1014276   G   A  Uncertain_significance   \n",
       "10  ISG15:9636  542076      1  1014401   G   A           Likely_benign   \n",
       "11  ISG15:9636  475281      1  1014451   C   T                  Benign   \n",
       "12  ISG15:9636  475282      1  1014471   G   C           Likely_benign   \n",
       "\n",
       "                                                CLNDN  \n",
       "0   Immunodeficiency_38_with_basal_ganglia_calcifi...  \n",
       "1   Immunodeficiency_38_with_basal_ganglia_calcifi...  \n",
       "3   Immunodeficiency_38_with_basal_ganglia_calcifi...  \n",
       "4   Immunodeficiency_38_with_basal_ganglia_calcifi...  \n",
       "5                                       not_specified  \n",
       "6   Immunodeficiency_38_with_basal_ganglia_calcifi...  \n",
       "7   Immunodeficiency_38_with_basal_ganglia_calcifi...  \n",
       "10  Immunodeficiency_38_with_basal_ganglia_calcifi...  \n",
       "11  Immunodeficiency_38_with_basal_ganglia_calcifi...  \n",
       "12  Immunodeficiency_38_with_basal_ganglia_calcifi...  "
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# I want to peek at the other columns for non-dangerous mutations. \n",
    "\n",
    "clin_v2[~(clin_v2.CLNSIG.str.contains(\"Pathogenic\", case=False) \n",
    "        & ~clin_v2.CLNSIG.str.contains(\"Conflicting_interpretations\", case=False))\n",
    "       ].head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I understand that there are some uncertain significance values or risk factor-related values. For now, I will assume that only the ones coded as Pathogenic or Likely pathogenic as dangerous and relevant for my boss's knowledge.\n",
    "\n",
    "Also the POS IDs that contained non-numerics - they just happen to not meet my definition of \"dangerous\" so I don't have to worry about converting this into a numeric value. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1) The deliverables are the final result as a dataframe with a short discussion of any specifics. \n",
    "# (that is, what data you would present to your boss with the explanation of your results)\n",
    "\n",
    "# Dataframes \n",
    "result_full = clin_v2[(clin_v2.CLNSIG.str.contains(\"Pathogenic\", case=False) \n",
    "                       & ~clin_v2.CLNSIG.str.contains(\"Conflicting_interpretations\", case=False))\n",
    "                     ]\n",
    "\n",
    "result_short = clin_v2[(clin_v2.CLNSIG.str.contains(\"Pathogenic\", case=False) \n",
    "                        & ~clin_v2.CLNSIG.str.contains(\"Conflicting_interpretations\", case=False))\n",
    "                       ].head(100)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Full result shape: (19474, 8)\n",
      "Top 100 result shape: (100, 8)\n"
     ]
    }
   ],
   "source": [
    "# Limit your output to the first 100 harmful mutations \n",
    "# and tell your boss how many total harmful mutations were found in the file\n",
    "\n",
    "print(\"Full result shape:\", result_full.shape)\n",
    "\n",
    "print(\"Top 100 result shape:\", result_short.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "# Email to the boss\n",
    "\n",
    "Hey Boss,\n",
    "\n",
    "I analyzed the clinvar data and was able to identify some dangerous genetic mutations. \n",
    "\n",
    "Please find attached a pandas dataframe called `result_short`. It contains the top 100 rows from the full dataset, which is in the dataframe called `result_full`. \n",
    "\n",
    "There were a total of 19,474 mutations that I think are dangerous. Having zero backbround in biology, I defined \"dangerous\" as any mutation marked as either pathogenic or likely pathogenic in the dataset's additional information (`INFO`), under its clinical significance field (`CLNSIG`). I excluded any mutations with clinical significance that was either uncertain or having conflicting interpretations. Let me know if you want me to include them.\n",
    "\n",
    "As you requested, `result_short` should have the following fields:\n",
    "\n",
    "|Field name|Description|Comments|\n",
    "|---|---|---|\n",
    "|GENEINFO|Gene name  ||\n",
    "|ID|Mutation ID number  ||\n",
    "|CHROM|Mutation Position (chromosome & position)  ||\n",
    "|POS|Mutation Position (chromosome & position)  ||\n",
    "|REF|Mutation value (reference & alternate bases)  |Reference base|\n",
    "|ALT|Mutation value (reference & alternate bases)  |Alternate base|\n",
    "|CLNSIG|Clinical severity  |Filtered for pathogenic, excluding uncertain or conflicting interpretations|\n",
    "|CLNDN|Disease that is implicated|Not code, but English names you can read.|\n",
    "\n",
    "See below for the table output of the top 100 rows. Please let me know if you have any questions. Thanks!\n",
    "\n",
    "Best,\n",
    "Haerang\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENEINFO</th>\n",
       "      <th>ID</th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>CLNSIG</th>\n",
       "      <th>CLNDN</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>183381</td>\n",
       "      <td>1</td>\n",
       "      <td>1014143</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>161455</td>\n",
       "      <td>1</td>\n",
       "      <td>1014316</td>\n",
       "      <td>C</td>\n",
       "      <td>CG</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>161454</td>\n",
       "      <td>1</td>\n",
       "      <td>1014359</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>AGRN:375790</td>\n",
       "      <td>243036</td>\n",
       "      <td>1</td>\n",
       "      <td>1022225</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Congenital_myasthenic_syndrome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>AGRN:375790</td>\n",
       "      <td>243037</td>\n",
       "      <td>1</td>\n",
       "      <td>1022313</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Congenital_myasthenic_syndrome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>AGRN:375790</td>\n",
       "      <td>574478</td>\n",
       "      <td>1</td>\n",
       "      <td>1041354</td>\n",
       "      <td>CGCCCGCCAGGAGAATGTCTTCAAGAAGTTCGACG</td>\n",
       "      <td>C</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Myasthenic_syndrome,_congenital,_8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>AGRN:375790</td>\n",
       "      <td>126556</td>\n",
       "      <td>1</td>\n",
       "      <td>1041582</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Congenital_myasthenic_syndrome|Myasthenic_synd...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>AGRN:375790</td>\n",
       "      <td>243038</td>\n",
       "      <td>1</td>\n",
       "      <td>1042136</td>\n",
       "      <td>T</td>\n",
       "      <td>TC</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Congenital_myasthenic_syndrome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>237</th>\n",
       "      <td>AGRN:375790</td>\n",
       "      <td>489335</td>\n",
       "      <td>1</td>\n",
       "      <td>1049672</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Likely_pathogenic</td>\n",
       "      <td>Not_Given</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>270</th>\n",
       "      <td>AGRN:375790</td>\n",
       "      <td>243039</td>\n",
       "      <td>1</td>\n",
       "      <td>1050473</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Congenital_myasthenic_syndrome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>280</th>\n",
       "      <td>AGRN:375790</td>\n",
       "      <td>18241</td>\n",
       "      <td>1</td>\n",
       "      <td>1050575</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Congenital_myasthenic_syndrome|Myasthenic_synd...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>282</th>\n",
       "      <td>AGRN:375790</td>\n",
       "      <td>126555</td>\n",
       "      <td>1</td>\n",
       "      <td>1050763</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Congenital_myasthenic_syndrome|Myasthenic_synd...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>292</th>\n",
       "      <td>AGRN:375790</td>\n",
       "      <td>503670</td>\n",
       "      <td>1</td>\n",
       "      <td>1051309</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>Likely_pathogenic</td>\n",
       "      <td>Not_Given</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>314</th>\n",
       "      <td>AGRN:375790</td>\n",
       "      <td>243040</td>\n",
       "      <td>1</td>\n",
       "      <td>1051775</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Congenital_myasthenic_syndrome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>365</th>\n",
       "      <td>TNFRSF4:7293</td>\n",
       "      <td>96692</td>\n",
       "      <td>1</td>\n",
       "      <td>1213738</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Immunodeficiency_16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>371</th>\n",
       "      <td>B3GALT6:126792</td>\n",
       "      <td>60484</td>\n",
       "      <td>1</td>\n",
       "      <td>1232279</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Spondyloepimetaphyseal_dysplasia_with_joint_la...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>375</th>\n",
       "      <td>B3GALT6:126792</td>\n",
       "      <td>60493</td>\n",
       "      <td>1</td>\n",
       "      <td>1232294</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Ehlers-Danlos_syndrome,_progeroid_type,_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>387</th>\n",
       "      <td>B3GALT6:126792</td>\n",
       "      <td>60488</td>\n",
       "      <td>1</td>\n",
       "      <td>1232471</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Spondyloepimetaphyseal_dysplasia_with_joint_la...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>388</th>\n",
       "      <td>B3GALT6:126792</td>\n",
       "      <td>60489</td>\n",
       "      <td>1</td>\n",
       "      <td>1232478</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Spondyloepimetaphyseal_dysplasia_with_joint_la...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>390</th>\n",
       "      <td>B3GALT6:126792</td>\n",
       "      <td>452846</td>\n",
       "      <td>1</td>\n",
       "      <td>1232517</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Likely_pathogenic</td>\n",
       "      <td>Not_Given</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>391</th>\n",
       "      <td>B3GALT6:126792</td>\n",
       "      <td>60497</td>\n",
       "      <td>1</td>\n",
       "      <td>1232594</td>\n",
       "      <td>CGGCGCGCCCTGGAGCGGGAGCA</td>\n",
       "      <td>C</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Spondyloepimetaphyseal_dysplasia_with_joint_la...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>393</th>\n",
       "      <td>B3GALT6:126792</td>\n",
       "      <td>60490</td>\n",
       "      <td>1</td>\n",
       "      <td>1232630</td>\n",
       "      <td>GA</td>\n",
       "      <td>G</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Ehlers-Danlos_syndrome,_progeroid_type,_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>398</th>\n",
       "      <td>B3GALT6:126792</td>\n",
       "      <td>60494</td>\n",
       "      <td>1</td>\n",
       "      <td>1232684</td>\n",
       "      <td>GTGCTGGCCA</td>\n",
       "      <td>G</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Ehlers-Danlos_syndrome,_progeroid_type,_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>401</th>\n",
       "      <td>B3GALT6:126792</td>\n",
       "      <td>60486</td>\n",
       "      <td>1</td>\n",
       "      <td>1232744</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Spondyloepimetaphyseal_dysplasia_with_joint_la...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>406</th>\n",
       "      <td>B3GALT6:126792</td>\n",
       "      <td>522415</td>\n",
       "      <td>1</td>\n",
       "      <td>1232834</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Spondyloepimetaphyseal_dysplasia_with_joint_la...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>407</th>\n",
       "      <td>B3GALT6:126792</td>\n",
       "      <td>427130</td>\n",
       "      <td>1</td>\n",
       "      <td>1232859</td>\n",
       "      <td>C</td>\n",
       "      <td>CG</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Not_Given</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408</th>\n",
       "      <td>B3GALT6:126792</td>\n",
       "      <td>60492</td>\n",
       "      <td>1</td>\n",
       "      <td>1232859</td>\n",
       "      <td>CG</td>\n",
       "      <td>C</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Ehlers-Danlos_syndrome,_progeroid_type,_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>414</th>\n",
       "      <td>B3GALT6:126792</td>\n",
       "      <td>60495</td>\n",
       "      <td>1</td>\n",
       "      <td>1232897</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Spondyloepimetaphyseal_dysplasia_with_joint_la...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>416</th>\n",
       "      <td>B3GALT6:126792</td>\n",
       "      <td>60496</td>\n",
       "      <td>1</td>\n",
       "      <td>1232927</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Spondyloepimetaphyseal_dysplasia_with_joint_la...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>418</th>\n",
       "      <td>B3GALT6:126792</td>\n",
       "      <td>60485</td>\n",
       "      <td>1</td>\n",
       "      <td>1232972</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Spondyloepimetaphyseal_dysplasia_with_joint_la...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>493</th>\n",
       "      <td>GNB1:2782</td>\n",
       "      <td>224714</td>\n",
       "      <td>1</td>\n",
       "      <td>1806509</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Neurodevelopmental_Disability|hypotonia|Strabi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>494</th>\n",
       "      <td>GNB1:2782</td>\n",
       "      <td>431082</td>\n",
       "      <td>1</td>\n",
       "      <td>1806512</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>Likely_pathogenic</td>\n",
       "      <td>Mental_retardation,_autosomal_dominant_42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>495</th>\n",
       "      <td>GNB1:2782</td>\n",
       "      <td>224713</td>\n",
       "      <td>1</td>\n",
       "      <td>1806513</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Likely_pathogenic</td>\n",
       "      <td>Neurodevelopmental_Disability|hypotonia|Seizur...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>496</th>\n",
       "      <td>GNB1:2782</td>\n",
       "      <td>224712</td>\n",
       "      <td>1</td>\n",
       "      <td>1806514</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>Pathogenic/Likely_pathogenic</td>\n",
       "      <td>Neurodevelopmental_Disability|hypotonia|Strabi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>497</th>\n",
       "      <td>GNB1:2782</td>\n",
       "      <td>224711</td>\n",
       "      <td>1</td>\n",
       "      <td>1806515</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>Pathogenic/Likely_pathogenic</td>\n",
       "      <td>Neurodevelopmental_Disability|hypotonia|Intell...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>498</th>\n",
       "      <td>GNB1:2782</td>\n",
       "      <td>617622</td>\n",
       "      <td>1</td>\n",
       "      <td>1815801</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Likely_pathogenic</td>\n",
       "      <td>Mental_retardation,_autosomal_dominant_42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>568</th>\n",
       "      <td>SKI:6497</td>\n",
       "      <td>224869</td>\n",
       "      <td>1</td>\n",
       "      <td>2228828</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Shprintzen-Goldberg_syndrome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>569</th>\n",
       "      <td>SKI:6497</td>\n",
       "      <td>570752</td>\n",
       "      <td>1</td>\n",
       "      <td>2228834</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>Likely_pathogenic</td>\n",
       "      <td>Shprintzen-Goldberg_syndrome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>571</th>\n",
       "      <td>SKI:6497</td>\n",
       "      <td>37259</td>\n",
       "      <td>1</td>\n",
       "      <td>2228860</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Shprintzen-Goldberg_syndrome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>573</th>\n",
       "      <td>SKI:6497</td>\n",
       "      <td>37261</td>\n",
       "      <td>1</td>\n",
       "      <td>2228866</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic/Likely_pathogenic</td>\n",
       "      <td>Inborn_genetic_diseases|Shprintzen-Goldberg_sy...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>574</th>\n",
       "      <td>SKI:6497</td>\n",
       "      <td>37262</td>\n",
       "      <td>1</td>\n",
       "      <td>2228866</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Inborn_genetic_diseases|Shprintzen-Goldberg_sy...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>575</th>\n",
       "      <td>SKI:6497</td>\n",
       "      <td>37260</td>\n",
       "      <td>1</td>\n",
       "      <td>2228867</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Shprintzen-Goldberg_syndrome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>576</th>\n",
       "      <td>SKI:6497</td>\n",
       "      <td>39783</td>\n",
       "      <td>1</td>\n",
       "      <td>2228867</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Shprintzen-Goldberg_syndrome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>578</th>\n",
       "      <td>SKI:6497</td>\n",
       "      <td>39785</td>\n",
       "      <td>1</td>\n",
       "      <td>2228870</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Shprintzen-Goldberg_syndrome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>599</th>\n",
       "      <td>SKI:6497</td>\n",
       "      <td>37263</td>\n",
       "      <td>1</td>\n",
       "      <td>2229045</td>\n",
       "      <td>GTCCGACCGC</td>\n",
       "      <td>G</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Shprintzen-Goldberg_syndrome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>600</th>\n",
       "      <td>SKI:6497</td>\n",
       "      <td>39784</td>\n",
       "      <td>1</td>\n",
       "      <td>2229045</td>\n",
       "      <td>GTCCGACCGCTCC</td>\n",
       "      <td>G</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Shprintzen-Goldberg_syndrome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>605</th>\n",
       "      <td>SKI:6497</td>\n",
       "      <td>37258</td>\n",
       "      <td>1</td>\n",
       "      <td>2229113</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Shprintzen-Goldberg_syndrome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>606</th>\n",
       "      <td>SKI:6497</td>\n",
       "      <td>224868</td>\n",
       "      <td>1</td>\n",
       "      <td>2229115</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>Pathogenic/Likely_pathogenic</td>\n",
       "      <td>Shprintzen-Goldberg_syndrome|not_provided</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>607</th>\n",
       "      <td>SKI:6497</td>\n",
       "      <td>222819</td>\n",
       "      <td>1</td>\n",
       "      <td>2229118</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Likely_pathogenic</td>\n",
       "      <td>Shprintzen-Goldberg_syndrome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>622</th>\n",
       "      <td>SKI:6497</td>\n",
       "      <td>213703</td>\n",
       "      <td>1</td>\n",
       "      <td>2229265</td>\n",
       "      <td>CT</td>\n",
       "      <td>GG</td>\n",
       "      <td>Likely_pathogenic</td>\n",
       "      <td>Not_Given</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>626</th>\n",
       "      <td>SKI:6497</td>\n",
       "      <td>495053</td>\n",
       "      <td>1</td>\n",
       "      <td>2229305</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>Likely_pathogenic</td>\n",
       "      <td>Disproportionate_tall_stature</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>634</th>\n",
       "      <td>SKI:6497</td>\n",
       "      <td>463409</td>\n",
       "      <td>1</td>\n",
       "      <td>2229401</td>\n",
       "      <td>AGCTCAGCGAGCGCAGCGTCCGCGTGTACCACGAGTGCTTCGGCAA...</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Shprintzen-Goldberg_syndrome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>846</th>\n",
       "      <td>PEX10:5192</td>\n",
       "      <td>554342</td>\n",
       "      <td>1</td>\n",
       "      <td>2406483</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>Likely_pathogenic</td>\n",
       "      <td>Peroxisome_biogenesis_disorder_6A|Peroxisome_b...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>847</th>\n",
       "      <td>PEX10:5192</td>\n",
       "      <td>557020</td>\n",
       "      <td>1</td>\n",
       "      <td>2406483</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Likely_pathogenic</td>\n",
       "      <td>Peroxisome_biogenesis_disorder_6A|Peroxisome_b...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>850</th>\n",
       "      <td>PEX10:5192</td>\n",
       "      <td>6771</td>\n",
       "      <td>1</td>\n",
       "      <td>2406526</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Peroxisome_biogenesis_disorder_6B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>852</th>\n",
       "      <td>PEX10:5192</td>\n",
       "      <td>197886</td>\n",
       "      <td>1</td>\n",
       "      <td>2406528</td>\n",
       "      <td>G</td>\n",
       "      <td>GC</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Not_Given</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>857</th>\n",
       "      <td>PEX10:5192</td>\n",
       "      <td>282334</td>\n",
       "      <td>1</td>\n",
       "      <td>2406561</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Not_Given</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>862</th>\n",
       "      <td>PEX10:5192</td>\n",
       "      <td>555297</td>\n",
       "      <td>1</td>\n",
       "      <td>2406580</td>\n",
       "      <td>CA</td>\n",
       "      <td>C</td>\n",
       "      <td>Likely_pathogenic</td>\n",
       "      <td>Peroxisome_biogenesis_disorder_6A|Peroxisome_b...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>863</th>\n",
       "      <td>PEX10:5192</td>\n",
       "      <td>296273</td>\n",
       "      <td>1</td>\n",
       "      <td>2406580</td>\n",
       "      <td>CAG</td>\n",
       "      <td>C</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Zellweger_syndrome|Peroxisome_biogenesis_disor...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>866</th>\n",
       "      <td>PEX10:5192</td>\n",
       "      <td>553074</td>\n",
       "      <td>1</td>\n",
       "      <td>2406599</td>\n",
       "      <td>GCT</td>\n",
       "      <td>G</td>\n",
       "      <td>Likely_pathogenic</td>\n",
       "      <td>Peroxisome_biogenesis_disorder_6A|Peroxisome_b...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           GENEINFO      ID  CHROM      POS  \\\n",
       "2        ISG15:9636  183381      1  1014143   \n",
       "8        ISG15:9636  161455      1  1014316   \n",
       "9        ISG15:9636  161454      1  1014359   \n",
       "24      AGRN:375790  243036      1  1022225   \n",
       "26      AGRN:375790  243037      1  1022313   \n",
       "46      AGRN:375790  574478      1  1041354   \n",
       "49      AGRN:375790  126556      1  1041582   \n",
       "63      AGRN:375790  243038      1  1042136   \n",
       "237     AGRN:375790  489335      1  1049672   \n",
       "270     AGRN:375790  243039      1  1050473   \n",
       "280     AGRN:375790   18241      1  1050575   \n",
       "282     AGRN:375790  126555      1  1050763   \n",
       "292     AGRN:375790  503670      1  1051309   \n",
       "314     AGRN:375790  243040      1  1051775   \n",
       "365    TNFRSF4:7293   96692      1  1213738   \n",
       "371  B3GALT6:126792   60484      1  1232279   \n",
       "375  B3GALT6:126792   60493      1  1232294   \n",
       "387  B3GALT6:126792   60488      1  1232471   \n",
       "388  B3GALT6:126792   60489      1  1232478   \n",
       "390  B3GALT6:126792  452846      1  1232517   \n",
       "391  B3GALT6:126792   60497      1  1232594   \n",
       "393  B3GALT6:126792   60490      1  1232630   \n",
       "398  B3GALT6:126792   60494      1  1232684   \n",
       "401  B3GALT6:126792   60486      1  1232744   \n",
       "406  B3GALT6:126792  522415      1  1232834   \n",
       "407  B3GALT6:126792  427130      1  1232859   \n",
       "408  B3GALT6:126792   60492      1  1232859   \n",
       "414  B3GALT6:126792   60495      1  1232897   \n",
       "416  B3GALT6:126792   60496      1  1232927   \n",
       "418  B3GALT6:126792   60485      1  1232972   \n",
       "..              ...     ...    ...      ...   \n",
       "493       GNB1:2782  224714      1  1806509   \n",
       "494       GNB1:2782  431082      1  1806512   \n",
       "495       GNB1:2782  224713      1  1806513   \n",
       "496       GNB1:2782  224712      1  1806514   \n",
       "497       GNB1:2782  224711      1  1806515   \n",
       "498       GNB1:2782  617622      1  1815801   \n",
       "568        SKI:6497  224869      1  2228828   \n",
       "569        SKI:6497  570752      1  2228834   \n",
       "571        SKI:6497   37259      1  2228860   \n",
       "573        SKI:6497   37261      1  2228866   \n",
       "574        SKI:6497   37262      1  2228866   \n",
       "575        SKI:6497   37260      1  2228867   \n",
       "576        SKI:6497   39783      1  2228867   \n",
       "578        SKI:6497   39785      1  2228870   \n",
       "599        SKI:6497   37263      1  2229045   \n",
       "600        SKI:6497   39784      1  2229045   \n",
       "605        SKI:6497   37258      1  2229113   \n",
       "606        SKI:6497  224868      1  2229115   \n",
       "607        SKI:6497  222819      1  2229118   \n",
       "622        SKI:6497  213703      1  2229265   \n",
       "626        SKI:6497  495053      1  2229305   \n",
       "634        SKI:6497  463409      1  2229401   \n",
       "846      PEX10:5192  554342      1  2406483   \n",
       "847      PEX10:5192  557020      1  2406483   \n",
       "850      PEX10:5192    6771      1  2406526   \n",
       "852      PEX10:5192  197886      1  2406528   \n",
       "857      PEX10:5192  282334      1  2406561   \n",
       "862      PEX10:5192  555297      1  2406580   \n",
       "863      PEX10:5192  296273      1  2406580   \n",
       "866      PEX10:5192  553074      1  2406599   \n",
       "\n",
       "                                                   REF ALT  \\\n",
       "2                                                    C   T   \n",
       "8                                                    C  CG   \n",
       "9                                                    G   T   \n",
       "24                                                   G   A   \n",
       "26                                                   A   T   \n",
       "46                 CGCCCGCCAGGAGAATGTCTTCAAGAAGTTCGACG   C   \n",
       "49                                                   C   T   \n",
       "63                                                   T  TC   \n",
       "237                                                  C   T   \n",
       "270                                                  G   A   \n",
       "280                                                  G   C   \n",
       "282                                                  G   T   \n",
       "292                                                  C  CT   \n",
       "314                                                  G   A   \n",
       "365                                                  G   A   \n",
       "371                                                  A   G   \n",
       "375                                                  C   T   \n",
       "387                                                  A   G   \n",
       "388                                                  C   T   \n",
       "390                                                  G   A   \n",
       "391                            CGGCGCGCCCTGGAGCGGGAGCA   C   \n",
       "393                                                 GA   G   \n",
       "398                                         GTGCTGGCCA   G   \n",
       "401                                                  G   A   \n",
       "406                                                  T   C   \n",
       "407                                                  C  CG   \n",
       "408                                                 CG   C   \n",
       "414                                                  G   C   \n",
       "416                                                  G   A   \n",
       "418                                                  C   T   \n",
       "..                                                 ...  ..   \n",
       "493                                                  T   C   \n",
       "494                                                  C   G   \n",
       "495                                                  C   T   \n",
       "496                                                  A   C   \n",
       "497                                                  T   C   \n",
       "498                                                  C   T   \n",
       "568                                                  T   G   \n",
       "569                                                  A   C   \n",
       "571                                                  C   G   \n",
       "573                                                  G   A   \n",
       "574                                                  G   T   \n",
       "575                                                  G   A   \n",
       "576                                                  G   T   \n",
       "578                                                  C   A   \n",
       "599                                         GTCCGACCGC   G   \n",
       "600                                      GTCCGACCGCTCC   G   \n",
       "605                                                  G   A   \n",
       "606                                                  G   C   \n",
       "607                                                  G   A   \n",
       "622                                                 CT  GG   \n",
       "626                                                  C   A   \n",
       "634  AGCTCAGCGAGCGCAGCGTCCGCGTGTACCACGAGTGCTTCGGCAA...   A   \n",
       "846                                                  C   G   \n",
       "847                                                  C   T   \n",
       "850                                                  G   C   \n",
       "852                                                  G  GC   \n",
       "857                                                  C   A   \n",
       "862                                                 CA   C   \n",
       "863                                                CAG   C   \n",
       "866                                                GCT   G   \n",
       "\n",
       "                           CLNSIG  \\\n",
       "2                      Pathogenic   \n",
       "8                      Pathogenic   \n",
       "9                      Pathogenic   \n",
       "24                     Pathogenic   \n",
       "26                     Pathogenic   \n",
       "46                     Pathogenic   \n",
       "49                     Pathogenic   \n",
       "63                     Pathogenic   \n",
       "237             Likely_pathogenic   \n",
       "270                    Pathogenic   \n",
       "280                    Pathogenic   \n",
       "282                    Pathogenic   \n",
       "292             Likely_pathogenic   \n",
       "314                    Pathogenic   \n",
       "365                    Pathogenic   \n",
       "371                    Pathogenic   \n",
       "375                    Pathogenic   \n",
       "387                    Pathogenic   \n",
       "388                    Pathogenic   \n",
       "390             Likely_pathogenic   \n",
       "391                    Pathogenic   \n",
       "393                    Pathogenic   \n",
       "398                    Pathogenic   \n",
       "401                    Pathogenic   \n",
       "406                    Pathogenic   \n",
       "407                    Pathogenic   \n",
       "408                    Pathogenic   \n",
       "414                    Pathogenic   \n",
       "416                    Pathogenic   \n",
       "418                    Pathogenic   \n",
       "..                            ...   \n",
       "493                    Pathogenic   \n",
       "494             Likely_pathogenic   \n",
       "495             Likely_pathogenic   \n",
       "496  Pathogenic/Likely_pathogenic   \n",
       "497  Pathogenic/Likely_pathogenic   \n",
       "498             Likely_pathogenic   \n",
       "568                    Pathogenic   \n",
       "569             Likely_pathogenic   \n",
       "571                    Pathogenic   \n",
       "573  Pathogenic/Likely_pathogenic   \n",
       "574                    Pathogenic   \n",
       "575                    Pathogenic   \n",
       "576                    Pathogenic   \n",
       "578                    Pathogenic   \n",
       "599                    Pathogenic   \n",
       "600                    Pathogenic   \n",
       "605                    Pathogenic   \n",
       "606  Pathogenic/Likely_pathogenic   \n",
       "607             Likely_pathogenic   \n",
       "622             Likely_pathogenic   \n",
       "626             Likely_pathogenic   \n",
       "634                    Pathogenic   \n",
       "846             Likely_pathogenic   \n",
       "847             Likely_pathogenic   \n",
       "850                    Pathogenic   \n",
       "852                    Pathogenic   \n",
       "857                    Pathogenic   \n",
       "862             Likely_pathogenic   \n",
       "863                    Pathogenic   \n",
       "866             Likely_pathogenic   \n",
       "\n",
       "                                                 CLNDN  \n",
       "2    Immunodeficiency_38_with_basal_ganglia_calcifi...  \n",
       "8    Immunodeficiency_38_with_basal_ganglia_calcifi...  \n",
       "9    Immunodeficiency_38_with_basal_ganglia_calcifi...  \n",
       "24                      Congenital_myasthenic_syndrome  \n",
       "26                      Congenital_myasthenic_syndrome  \n",
       "46                  Myasthenic_syndrome,_congenital,_8  \n",
       "49   Congenital_myasthenic_syndrome|Myasthenic_synd...  \n",
       "63                      Congenital_myasthenic_syndrome  \n",
       "237                                          Not_Given  \n",
       "270                     Congenital_myasthenic_syndrome  \n",
       "280  Congenital_myasthenic_syndrome|Myasthenic_synd...  \n",
       "282  Congenital_myasthenic_syndrome|Myasthenic_synd...  \n",
       "292                                          Not_Given  \n",
       "314                     Congenital_myasthenic_syndrome  \n",
       "365                                Immunodeficiency_16  \n",
       "371  Spondyloepimetaphyseal_dysplasia_with_joint_la...  \n",
       "375          Ehlers-Danlos_syndrome,_progeroid_type,_2  \n",
       "387  Spondyloepimetaphyseal_dysplasia_with_joint_la...  \n",
       "388  Spondyloepimetaphyseal_dysplasia_with_joint_la...  \n",
       "390                                          Not_Given  \n",
       "391  Spondyloepimetaphyseal_dysplasia_with_joint_la...  \n",
       "393          Ehlers-Danlos_syndrome,_progeroid_type,_2  \n",
       "398          Ehlers-Danlos_syndrome,_progeroid_type,_2  \n",
       "401  Spondyloepimetaphyseal_dysplasia_with_joint_la...  \n",
       "406  Spondyloepimetaphyseal_dysplasia_with_joint_la...  \n",
       "407                                          Not_Given  \n",
       "408          Ehlers-Danlos_syndrome,_progeroid_type,_2  \n",
       "414  Spondyloepimetaphyseal_dysplasia_with_joint_la...  \n",
       "416  Spondyloepimetaphyseal_dysplasia_with_joint_la...  \n",
       "418  Spondyloepimetaphyseal_dysplasia_with_joint_la...  \n",
       "..                                                 ...  \n",
       "493  Neurodevelopmental_Disability|hypotonia|Strabi...  \n",
       "494          Mental_retardation,_autosomal_dominant_42  \n",
       "495  Neurodevelopmental_Disability|hypotonia|Seizur...  \n",
       "496  Neurodevelopmental_Disability|hypotonia|Strabi...  \n",
       "497  Neurodevelopmental_Disability|hypotonia|Intell...  \n",
       "498          Mental_retardation,_autosomal_dominant_42  \n",
       "568                       Shprintzen-Goldberg_syndrome  \n",
       "569                       Shprintzen-Goldberg_syndrome  \n",
       "571                       Shprintzen-Goldberg_syndrome  \n",
       "573  Inborn_genetic_diseases|Shprintzen-Goldberg_sy...  \n",
       "574  Inborn_genetic_diseases|Shprintzen-Goldberg_sy...  \n",
       "575                       Shprintzen-Goldberg_syndrome  \n",
       "576                       Shprintzen-Goldberg_syndrome  \n",
       "578                       Shprintzen-Goldberg_syndrome  \n",
       "599                       Shprintzen-Goldberg_syndrome  \n",
       "600                       Shprintzen-Goldberg_syndrome  \n",
       "605                       Shprintzen-Goldberg_syndrome  \n",
       "606          Shprintzen-Goldberg_syndrome|not_provided  \n",
       "607                       Shprintzen-Goldberg_syndrome  \n",
       "622                                          Not_Given  \n",
       "626                      Disproportionate_tall_stature  \n",
       "634                       Shprintzen-Goldberg_syndrome  \n",
       "846  Peroxisome_biogenesis_disorder_6A|Peroxisome_b...  \n",
       "847  Peroxisome_biogenesis_disorder_6A|Peroxisome_b...  \n",
       "850                  Peroxisome_biogenesis_disorder_6B  \n",
       "852                                          Not_Given  \n",
       "857                                          Not_Given  \n",
       "862  Peroxisome_biogenesis_disorder_6A|Peroxisome_b...  \n",
       "863  Zellweger_syndrome|Peroxisome_biogenesis_disor...  \n",
       "866  Peroxisome_biogenesis_disorder_6A|Peroxisome_b...  \n",
       "\n",
       "[100 rows x 8 columns]"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_short"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  },
  "latex_envs": {
   "LaTeX_envs_menu_present": true,
   "autoclose": false,
   "autocomplete": true,
   "bibliofile": "biblio.bib",
   "cite_by": "apalike",
   "current_citInitial": 1,
   "eqLabelWithNumbers": true,
   "eqNumInitial": 1,
   "hotkeys": {
    "equation": "Ctrl-E",
    "itemize": "Ctrl-I"
   },
   "labels_anchors": false,
   "latex_user_defs": false,
   "report_style_numbering": false,
   "user_envs_cfg": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
